pkg/services/ngalert/state/historian/dashboard.go

```diff
@@ -0,0 +1,89 @@
+package historian
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/patrickmn/go-cache"
+)
+
+const dashboardNotFound = int64(-1)
+
+const defaultDashboardCacheExpiry = 1 * time.Minute
+
+const minCleanupInterval = 1 * time.Second
+
+// dashboardResolver resolves dashboard UIDs to IDs with caching.
+type dashboardResolver struct {
+	dashboards dashboards.DashboardService
+	cache      *cache.Cache
+	lock       sync.Mutex
+	log        log.Logger
+}
```

Comment:
Possible TODO: metrics for cache hits and misses

Reply:
What about a periodic debug log? That would be easy to search without adding metrics, and at least in the alerting case, the number should be somewhat steady because it's driven by background work.

Reply:
Decided to hold off on this one for a bit. I didn't end up going the log route because the background work isn't regular, it can be very bursty. We can always add this in the future, especially if we use this utility in more than one place as George suggested.